### DISCUSSION TITLE
Takes a very long time to run

### DISCUSSION THREAD

**Discussion by Takodachi6969** [Part-II-Computational-Physics/.github#55] [General]
When I was running the program with the given plot times, it takes half an hour to run it once meaning there must be something wrong with my code. Upon testing smaller limits the code works similarly to how I expect it to behave. I am not so sure which part of it went wrong....

BTW is it ok to share code on this thread?

```
def init_lattice():
    '''Produce an initial lattice with spins. Type 1 gives all spins 1, type 0 gives a random spins, type -1 gives all spins -1. Captures invalid types.
    '''
    global width
    global type
    lattice = np.full((width, width), type)
    return lattice
    
    

def neighbouring_sites(i,j):
    '''Return the coordinates of the 4 sites adjacent to [i,j] on an width*width lattice. Takes into account periodic boundary conditions.
    
    '''
    global width
    Neighbour = []
    Neighbour.append([i, j+1])
    Neighbour.append([i+1, j])
    Neighbour.append([i, j-1])
    Neighbour.append([i-1, j])
    return np.array(Neighbour)
    

def neighbouring_spins_sum(i,j):
    '''Sums the spins of all neighbours of the spin at [i,j].
    '''
    global lattice
    k = np.asarray([[0,1,0], [1,0,1], [0,1,0]])
    U = scipy.signal.convolve2d(lattice, k)
    return U[i,j]

def compute_magnetisation():
    '''Computes the magnetisation of the lattice.
    '''
    global lattice
    global width
    N = width * width
    M = (1 / N ) * np.sum(lattice)
    return M

def plot_lattice(lattice,ax,title):
    '''Plot the lattice configuration.'''
    ax.matshow(lattice, vmin=-1, vmax=1, cmap=plt.cm.binary)
    ax.title.set_text(title)
    ax.set_yticklabels([])
    ax.set_xticklabels([])
    ax.set_yticks([])
    ax.set_xticks([])

def compute_betaDeltaE(i,j):
    '''Computes the energy difference between the old and new state if spin [i,j] would be flipped.'''
    global lattice
    global width
    global betaJ
    global betaMuH
    E = 2 * betaJ * lattice[i,j] * neighbouring_spins_sum(i,j) + 2 * betaMuH * lattice[i,j]
    return E

def attempt_spin_flip():
    '''Applies the Metropolis-Hastings algorithm to try and flip a spin.'''
    global width
    global lattice
    global betaJ
    global betaMuH
    
    for i in range(width):
        for j in range(width):
            if rng.random() <= np.exp(-1 * compute_betaDeltaE(i,j)):
                lattice[i,j] *= -1
    return lattice
        

def evolve_and_plot():
    '''Evolves the lattice using the Metropolis-Hastings algorithm and plots the lattice at different times.'''
    global betaJ
    global betaMuH
    global lattice
    global width
    global plot_times
  
    fig, ax = plt.subplots(1,len(plot_times),figsize=(20,4))
    for t in range(plot_times[-1]+1):
        lattice = attempt_spin_flip()
        if t in plot_times:
            plot_lattice(lattice,ax[plot_times.index(t)],"t = {}".format(t))
    plt.show()

def evolve_and_compute_M():
    '''Evolves the lattice using the Metropolis-Hastings algorithm and returns the average magnetisation computed using different time steps.'''
    global betaJ
    global betaMuH
    global lattice
    global width
    global plot_times
    global avg_times
    
    Mag = []
    
    for t in range(avg_times[-1]+1):
        lattice = attempt_spin_flip()
        if t in avg_times:
            Mag.append(compute_magnetisation())
    return Mag

plot_times = [0,10,100,1000]
width = 50
type = 1
lattice = init_lattice()
betaJ = 0
betaMuH = 3

evolve_and_plot()
```

**Comment by Jamiewhbee** [comment]
With a lattice size of 50, you're looking at 4x50x50 lattice sites, and then changing those lattice sites, at each of the 1000 time steps. Maybe a smaller lattice size (10 or 15?) would make it run in a more reasonable time.
Also having to call each of the 50x50 lattice sites as a global variable before doing every one of those time steps and energy calculations might be contributing to the run time? I'm not sure exactly what the computational process for global variables looks like, so maybe it doesn't actually affect it, but in the worst case, you might have multiplied the computational time by yet another factor of 50.

**Reply by Takodachi6969** [reply]
Its Python so global variable doesn't matter at all. Doing smaller lattice did a bit, but the run time is still too long

**Reply by Jamiewhbee** [reply]
Removing the global variables might reduce the run time further.

**Comment by Takodachi6969** [comment]
I think I found the mistake, instead of iterating every single cell in the attempt_spin_flip function I should just done it once and repeat for 10^5 times, I was doing 10^10 calculations which was why it was kinda slow...

But then this doesn't explain why we shouldn't do all cells as each individual cell has that value of flipping

**Reply by DLHvdH** [reply]
Hi, you are right. You should be doing a single flip attempt per time step for computational reasons. It is true that actually all spins should have a probability of flipping at a given point in time. However, you might also notice that we never decided on a time step. The current model thus doesn't really have a physical time, and whether we decide to flip a single spin or multiple spins in a step doesn't really matter. In our case we are only interested in equilibrium properties.
